KEP-5073: Declarative Validation: Add validation generator

jpbetz opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is the first of a series of PRs implementing Declarative Validation.

This PR introduces validation-gen, the generator that will produce declarative validation code based on the `//+` tags in `types.go` files.

Individual validation rules are "plugable". This PR introduces a small set of "core" validation plugins. Subsequent PRs will add more plugins to perform validations (for example: "dnsName", "enum", "union" and similar will be added later).

This PR includes test infrastructure for the generator and a large set of test cases demonstrating that correct validation is generated for a wide range of conditions.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

Release note will be added as part of subsequent PRs for this feature.

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

Implements the code generator for Declarative Validation

KEP: https://github.com/kubernetes/enhancements/issues/5073

jpbetz commented:
/sig api-machinery
/triage accepted
/cc @thockin @aaron-prindle @yongruilin

jpbetz commented:
/retest-required

Aaina26 commented:
/cc  ref #130358

deads2k commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go:
```diff
@@ -0,0 +1,210 @@
+
+	switch rtv.mode {
+	case requirednessRequired:
+		doc.Description = "Indicates that a field is optional to clients."
```
`requirednessRequired` why is this one optional to clients?  backwards comment?

aaron-prindle replied:
You are correct, the docs for requirednessRequired and requirednessOptional are currently flipped. I've updated the lines here so they properly describe the +k8s:required tag as indicating a required field, and +k8s:optional as indicating an optional field. Thanks for catching this!

aaron-prindle replied:
+1, I believe this initially was due to a cut-paste error. This is resolved now with https://github.com/kubernetes/kubernetes/pull/130349/commits/314819ef353ccfaa1311ce750039bba8458d7a3a

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/each.go:
```diff
@@ -0,0 +1,117 @@
+		if cmp != nil && len(oldSlice) > 0 {
+			old = lookup(oldSlice, val, cmp)
+		}
+		errs = append(errs, validator(opCtx, fldPath.Index(i), &val, old)...)
```
From an end-user POV, I think index is sufficient - the goal is to help them find the error, right?  With very few exceptions, the number of items in a list is usually pretty small.

jpbetz replied:
I've wondered about exactly this question in the past as well.  I agree with the observation the author of apply configuration does not have sufficient information to map the indices of a listType=map entries in validation errors back to the contents of their apply configuration.  OK if we track this separately?

deads2k replied:
> OK if we track this separately?

Yeah, that's fine.  Observation, not requirement.


> the goal is to help them find the error, right? With very few exceptions, the number of items in a list is usually pretty small.

While true, as a human knowing that it's `condition[@name=Available]` is more useful than `condition[2]`.  I don't mind a future consideration.

aaron-prindle replied:
Created a tracking bug for this here:
https://github.com/kubernetes/kubernetes/issues/130454

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/README.md:
```diff
@@ -0,0 +1,60 @@
+        "k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+func <Name>(opCtx operation.Context, fldPath *field.Path, value, oldValue <ValueType>, <OtherArgs...>) field.ErrorList
```
Oof.  We could, I guess.  I would beg permission to explorse that as a followup, since it will require surgery in the generator.

deads2k replied:
Without a context, both contexual logging and warnings are impossible.  Is it that much surgery in the generator to thread it to all calls?  If not now, when is more convenient.

thockin replied:
https://github.com/jpbetz/kubernetes/pull/99/commits/9bbd5335e1e4f16c52022a35f0b2a4825eb10b15

jpbetz replied:
Commits to add context are merged into this PR

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/lint.go:
```diff
@@ -0,0 +1,160 @@
+
+// ruleRequiredAndDefault checks for conflicting tags +k8s:required and +default in a slice of comments.
+func ruleRequiredAndDefault(comments []string) (string, error) {
+	return conflictingTagsRule(comments, "+k8s:required", "+default")
```
Discussion today was not complete - let's follow up tomorrow.

I want to use only new tags to trigger declarative validation (`+k8s:optional` vs `+optional`), but I want to aim for a result where that is 100% semantically equal to the old tag, so we eventually don't need both and can merge them.  If we find a place where we want `+optional` and `+k8s:required` on the same field, I think we took a wrong turn

The only exception (maybe) to this is that I currently peek at `+default` (not `+k8s:default`) because that tag does not __trigger__ any generation.  I could make that `+k8s:default` as long as we are OK that it is alway exactly the same as `+default` and that (as with optional) the goal is to merge them.

For me the distinction was whether it TRIGGERS code-generation or not.

deads2k replied:
Eventually resolved the thread.  We'll make this tool aware of tags from both built-ins and at least kubebuilder.

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/lint.go:
```diff
@@ -0,0 +1,160 @@
+	lintErrors []error
+}
+
+var defaultRules = []lintRule{
+	ruleOptionalAndRequired,
+	ruleRequiredAndDefault,
+}
```
We chose `+k8s:` as a short-but-unique prefix to avoid triggering code-gen on existing tags (e.g. `+optional`).  I don't LOVE it as a prefix, but I did not want `+k8s:validation:` or something long.

The question (opinion) I have for you is this:  Fast forward - we have converted everything we can convert and added dozens of new validation tags.  They all say `+k8s:this` and `+k8s:that`, but we still have a small number of tags like `+default` with no prefix.  Would you rather:  a) accept `+k8s:default` as an alias to `+default` (use the prefix everywhere) or b) accept `+format` as an alias for `+k8s:format` and `+minItems` as an alias for `+k8s:minItems` (effectively "get rid of the prefix")?

deads2k replied:
On balance, I prefer having `+k8s` as prefix. I was intending to ask whether a linter should fail on `+k8s:misspelled-tag`.  Pro: you can't typo something.  Con: you can't run old generators on new golang types.

yongruilin replied:
Adding lint check for `misspelled-tag` is a good idea but maybe in a future separate PR? For validation-gen, we have the interface `TagName()` to consolidate from the registry, and for other generators (conversion, deepcopy, etc.) I can only think of doing hardcoded to gather them together.

jpbetz replied:
> Pro: you can't typo something. Con: you can't run old generators on new golang types.

This is exactly what surfaced in my mind on this one.  The "con" here is what I'm snagged on.. I'm having trouble convincing myself that it's okay to break tools that don't understand new tags.

aaron-prindle replied:
Created a tracking bug for this here:
https://github.com/kubernetes/kubernetes/issues/130455

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/lint.go:
```diff
@@ -0,0 +1,160 @@
+	lintErrors []error
+}
+
+var defaultRules = []lintRule{
+	ruleOptionalAndRequired,
+	ruleRequiredAndDefault,
+}
```
That argument is a string whose job is to appear in the generated code and be verified in tests.  We can't easily verify the string's contents, but maybe we don't need them now that we have the validateFalse fixture test?

@jpbetz ?

deads2k replied:
To clarify, a problem I've seen is `+kubebuilder:validation:XValidation:rule="",misspelled-message="msg"`, where `misspelled-message=` does not cause a failure

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/operation/context.go:
```diff
@@ -0,0 +1,48 @@
+	// feature enabled even when the feature gate is disabled. This can happen when feature is
+	// already in-use by a resource, often because the feature gate was enabled when the
+	// resource first began using the feature.
+	Options sets.Set[string]
```
Yes, `unset == false`. I'll clarify.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/safe/safe.go:
```diff
@@ -0,0 +1,78 @@
+// Field takes a pointer to any value (which may or may not be nil) and
+// a function that traverses to a target type R (a typical use case is to dereference a field),
+// and returns the result of the traversal, or the zero value of the target type.
+// This is roughly equivalent to "value != nil ? fn(value) : nil" in languages that support the ternary operator.
```
Correct. Fixing godoc.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/safe/safe.go:
```diff
@@ -0,0 +1,78 @@
+
+// Lookup looks up the specified key in the map and, if found, applies a
+// transform function.  If the map is nil or the key is not found, it returns
+// the zero-value for the map's value type.  The transform function will only
+// be called if the map is non-nil and the key is found.
```
I did some searching and it turns out that `safe.Field` and `safe.Cast` are the only functions we still use from this package.

I propose we remove `Lookup`, `PtrTo`, `Deref` and `Ident`. Apologies for the cruft.

cc @thockin

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/safe/safe.go:
```diff
@@ -0,0 +1,78 @@
+	return xform(val)
+}
+
+// PtrTo returns a pointer to the argument value.  This works as a transform
+// function for Lookup, e.g. given a map with non-pointer values, this will
+// produce a pointer.
+func PtrTo[T any](val T) *T {
+	return &val
+}
+
+// Deref returns the value pointed to by the argument pointer, or the zero
+// value if the pointer is nil.  This works as a transform function for Lookup,
+// e.g. given a map with pointer-to-slice values, this will produce a slice.
+func Deref[T any](val *T) T {
+	if val == nil {
+		var zero T
+		return zero
+	}
+	return *val
+}
```
This is cruft. Dropping these functions. (xref https://github.com/kubernetes/kubernetes/pull/130349#discussion_r1970233981)

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/api/validate/README.md:
```diff
@@ -0,0 +1,60 @@
+
+The `value` and `oldValue` arguments are always nilable - pointers to primitive
+types, slices of any type, or maps of any type.  Validator functions should
+avoid dereferencing nil, but callers are expected to not pass nil unless the
+API field itself was nilable.
```
> if it's a create, I would expect `oldValue` to be nil.

Yes, this is what happens. 

> Is that not the case for non-pointers?

I see how the working here has caused confusion.  There is a statement further up along the lines of "the `oldValue` argument will be nil", but the statement that "callers are expected to not pass nil unless the
API field itself was nilable" seems to be saying something else.  I'll clarify.

thockin commented on staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go:
```diff
@@ -0,0 +1,131 @@
+
+// RequiredValue verifies that the specified value is not the zero-value for
+// its type.
+func RequiredValue[T comparable](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
```
I propose we rename this file to requiredness.go or required.go or something.

deads2k commented:
I'm happy with it except for the handling of types outside a package boundary (assuming I understood it correctly) https://github.com/kubernetes/kubernetes/pull/130349#discussion_r1975535748

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go:
```diff
@@ -0,0 +1,1402 @@
+				emitComments(validations.Comments, bufsw)
+			}
+
+			// If the node is nil, this must be a type in a package we are not
+			// handling - it's effectively opaque to us.
+			if fld.node == nil {
+				targs := generator.Args{
+					"type": fld.childType,
+				}
+				bufsw.Do("// NOTE: Not validating this field's type: $.type|raw$ is in a non-included package.\n", targs)
```
I like Option 2.

https://github.com/jpbetz/kubernetes/pull/109 could implement this.

WDYT @thockin?  We could refine this later if needed.

thockin commented:
I hoped maybe an extra commit to fix merge would be enough but it's obviously not :)  Not sure what I was thinking.  Joe, the rebase is trivial - 1 conflict:

```diff
diff --git a/vendor/modules.txt b/vendor/modules.txt
index 83cf5a982dd..2b39e46c829 100644
--- a/vendor/modules.txt
+++ b/vendor/modules.txt
@@ -1042,7 +1042,7 @@ gopkg.in/yaml.v3
 # k8s.io/endpointslice v0.0.0 => ./staging/src/k8s.io/endpointslice
 ## explicit; go 1.23.0
 # k8s.io/externaljwt v0.0.0 => ./staging/src/k8s.io/externaljwt
-## explicit; go 1.23.0
+## explicit; go 1.24.0
 # k8s.io/gengo/v2 v2.0.0-20250207200755-1244d31929d7
 ## explicit; go 1.20
 k8s.io/gengo/v2
```

jpbetz commented:
> I'm happy with it except for the handling of types outside a package boundary (assuming I understood it correctly) [#130349 (comment)](https://github.com/kubernetes/kubernetes/pull/130349#discussion_r1975535748)

The last few commits (same ones as in https://github.com/jpbetz/kubernetes/pull/111) implement Option 2.

The tag is called `opaqueType` and can also be used to skip validation on a imported type, which may come handy in situations where we have type sharing but the validation differs.  This can compose with `+subfield` (and similar) to "use the go type, but replace the validation".

jpbetz committed:
Add validation-gen to codegen scripts

Add validation-gen to repository wide codegen scripts and introduce
`./hack/update-codegen.sh validation` as a quick way to run validation-gen.

jpbetz committed:
Bump gengo/v2 to latest

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>
# Conflicts:
#	vendor/modules.txt

jpbetz committed:
Introduce validation-gen

Adds code-generator/cmd/validation-gen. This provides the machinery
to discover `//+` tags in types.go files, register plugins to handle
the tags, and generate validation code.

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

jpbetz committed:
Add validators: opaque

Skips validation of a type's tags.

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

jpbetz committed:
Add validators: eachkey, eachval, subfield

Introduce a composable set of tags for validating child data.
This allows for point-of-use validation of shared types.

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

jpbetz committed:
Add validators: optional/required/forbidden

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

jpbetz committed:
Add validation-gen test infrastructure

Introduces the infrastructure for testing validation-gen tags.

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

jpbetz committed:
Add validators: immutable

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

jpbetz committed:
Add validation-gen core tests

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

deads2k commented:
> The tag is called opaqueType and can also be used to skip validation on a imported type, which may come handy in situations where we have type sharing but the validation differs. This can compose with +subfield (and similar) to "use the go type, but replace the validation".

Interesting.  I was about to observe that I think it could be used "improperly" to skip validation on a type we do have.  Not sure I see the value in different validation, but I guess I wouldn't rule it out.

deads2k commented:
lgtm

deads2k commented:
/lgtm
/approve

jpbetz commented:
@dims or @liggitt Could we get one or your approvals for the gengo deps bump here? (https://github.com/kubernetes/kubernetes/pull/130349/commits/7f5e1baeeea7382e45cfba7676b69459fbce8b12)

thockin commented:
/retest
